File: Go/20.go
/**
 * Runtime: 1ms (Beats 71.23%)
 * Memory: 2.2MB (Beats 20.59%)
 */

package main

func isValid(s string) bool {
	stack := []string{}
	for _, char := range s {
		if char == '{' || char == '(' || char == '[' {
			stack = append(stack, string(char))
		} else {
			if len(stack) == 0 {
				return false
			} else if char == '}' && stack[len(stack)-1] == "{" {
				stack = stack[:len(stack)-1]
			} else if char == ')' && stack[len(stack)-1] == "(" {
				stack = stack[:len(stack)-1]
			} else if char == ']' && stack[len(stack)-1] == "[" {
				stack = stack[:len(stack)-1]
			} else {
				return false
			}
		}
	}
	return len(stack) == 0
}

File: JavaScript/src/20.js
/**
 * @param {string} s
 * @return {boolean}
 */
function isValid(s) {
  if (s.length % 2 !== 0) {
    return false;
  }

  const stack = [];
  const matchingBrackets = {
    "(": ")",
    "{": "}",
    "[": "]",
  };

  for (const char of s) {
    if (matchingBrackets[char]) {
      stack.push(matchingBrackets[char]);
    } else {
      if (stack.pop() !== char) {
        return false;
      }
    }
  }

  return stack.length === 0;
}

export { isValid };
